File: 2021/lib/day02.rb
# https://adventofcode.com/2021/day/2

class Day02

  def self.part1(input)
    position = 0
    depth = 0
    input.each do |line|
      /(?<command>forward|down|up) (?<amount>\d+)/ =~ line
      case command.to_sym
      when :forward
        position += amount.to_i
      when :up
        depth -= amount.to_i
      when :down
        depth += amount.to_i
      end
    end
    [position, depth]
  end

  def self.part2(input)
    position = 0
    aim = 0
    depth = 0
    input.each do |line|
      /(?<command>forward|down|up) (?<amount>\d+)/ =~ line
      case command.to_sym
      when :forward
        position += amount.to_i
        depth += aim * amount.to_i
      when :up
        aim -= amount.to_i
      when :down
        aim += amount.to_i
      end
    end
    [position, depth]
  end
end

File: 2021/lib/day02.exs
# Answer based on Jose Valim's stream
# I'm learning elixir

{_, input} = File.read("input/day02.txt") 

part1 = input
  |> String.split("\n", trim: true)
  |> Enum.reduce({_depth = 0, _position = 0}, fn
    "forward " <> number, {depth, position} -> {depth, position + String.to_integer(number)}
    "down " <> number, {depth, position} -> {depth + String.to_integer(number), position}
    "up " <> number, {depth, position} -> {depth - String.to_integer(number), position}
  end)
  |> then(fn {depth, position} -> depth * position end)

part2 = input
  |> String.split("\n", trim: true)
  |> Enum.reduce({_depth = 0, _position = 0, _aim = 0}, fn
    "forward " <> number, {depth, position, aim} ->
      number = String.to_integer(number)
      {depth + aim * number, position + number, aim}

    "down " <> number, {depth, position, aim} ->
      {depth, position, aim + String.to_integer(number)}

    "up " <> number, {depth, position, aim} ->
      {depth, position, aim - String.to_integer(number)}
  end)
  |> then(fn {depth, position, _aim} -> depth * position end)

defmodule Submarine do
  def navigate("forward " <> number, {depth, position, aim}) do
    number = String.to_integer(number)
    {depth + aim * number, position + number, aim}
  end

  def navigate("down " <> number, {depth, position, aim}) do
    {depth, position, aim + String.to_integer(number)}
  end

  def navigate("up " <> number, {depth, position, aim}) do
    {depth, position, aim - String.to_integer(number)}
  end
end

part2a = input
  |> String.split("\n", trim: true)
  |> Enum.reduce({_depth = 0, _position = 0, _aim = 0}, &Submarine.navigate/2)
  |> then(fn {depth, position, _aim} -> depth * position end)

IO.write("Part 1: ")
IO.puts(part1)
IO.write("Part 2: ")
IO.puts(part2)
IO.write("Part 2a:")
IO.puts(part2a)
